\chapter{Data}

InfectX is an \gls{rtd} project funded by SystemsX, the Swiss initiative for systems biology with the goal of studying and identifying components of the human infectome for a set of bacterial and viral pathogens. A central effort consists of generating kinome- and genome-wide \gls{sirna} screens for each of the investigated pathogens, capturing image data and performing image analysis. The acquired data is publicly available on the internet at \href{http://www.infectx.ch/databrowser}{the InfectX website}.

\section{InfectX Workflow}
Due to the large scale accompanied by screening multiple libraries, using numerous pathogens and desiring replicates, several labs were involved in carrying out wet-lab experimentation. In order to obtain reproducible results, a strong emphasis was put on developing unified procedures for cell culture, \gls{sirna} transfection, pathogen infection and imaging. Figure \ref{fig:infectx} summarizes the central steps, beginning with \gls{sirna} libraries arrayed onto 384-well plates that are used for transfection of added cells, carrying on with pathogen infection, subsequent microscopic imaging and concluding with computational image analysis. Much of the information contained in this chapter has been published by \cite{Ramo2014} and is summarized for the reader's convenience.

\begin{figure}
  \centering
  \includegraphics[width=0.95\textwidth]{infectx}
  \caption[InfectX RNAi data acquisition and analysis workflow.]{A total of 11 single \gls{sirna} libraries, pdroduced by 3 separate manufacturers, (4 from Dharmacon, 4 from Quiagen and 3 from Ambion), as well as one pooled library (Dharmacon) were screened with 8 pathogens. Plates were imaged under wide-field microscopes and the resulting images run through an image analysis pipeline. \citep{Ramo2014}.}
  \label{fig:infectx}
\end{figure}

The model system chosen for investigation is HeLa (ATCC CCL-2), the oldest and most wide spread human cell line and a proven system for studying infelicitous disease\footnote{Collected 60 years ago from a cervical adenocarcinoma, these epithelial cells have led to much insight into human cell biology. Prior to their discovery, attempts at growing human tissue in vitro were futile and development of protocols for sustaining cancerous human tissue was thought to hold great promise for cancer research. One of the early successes involving HeLa cells was the development a polio vaccine. For this endeavor, a large amount of human cells were needed and the installment of a production facility capable of meeting the high demand might have contributed significantly to the predominance of these cells. \citep{Masters2002}}. Culturing was performed at \SI{37}{\celsius}, under 5\% \ce{CO2} atmosphere for maintaining optimal pH, using \gls{dmem}  supplemented with 10\% inactivated \gls{fbs}, both supplied by Invitrogen.

While genome-wide \gls{sirna} screens were also produced within the InfectX framework, this report focuses on kinome-wide investigations. Introduced by \cite{Manning2002}, the term kinome refers to the subset of genes encoding protein kinases\footnote{Kinases are part of the larger enzymatic family of transferases and catalyze phosphorylation reactions of the form
\begin{equation}
  \ce{A\bond{-}O\bond{-}PO3^{-2} + H\bond{-}X\bond{-}B <=>> B\bond{-}X\bond{-}O\bond{-}PO3^{-2} + H\bond{-}O\bond{-}A} \label{eq:kinase}
\end{equation}
where A represents the donating (typically \gls{atp}) and B the accepting molecule. Kinases are further subdivided according to the type of acceptor which can be an alcohol, carboxy, nitrogenous or phosphate group, or in case of protein kinases, a tyrosine, serine, threonine or histidine residue.
\begin{align}
  \cee{A\bond{-}O\bond{-}B + H\bond{-}O\bond{-}PO3^{-2} &<=>> A\bond{-}O\bond{-}PO3^{-2} + H\bond{-}O\bond{-}B} \label{eq:phosphorylase} \\
  \cee{A\bond{-}O\bond{-}PO3^{-2} + H\bond{-}OH &<=>> A\bond{-}OH + H\bond{-}O\bond{-}PO3^{-2}} \label{eq:phosphatases}
\end{align}
Phosphorylases are a further group of transferases that involve phosphate but unlike kinases, they utilize inorganic phosphate sources (\ref{eq:phosphorylase}). Often phosphorylases are involved in breaking down biological polymers such as polysaccharides and polynucleotides. Finally, phosphatases catalyze the reverse reaction, the removal of a phosphate group (\ref{eq:phosphatases}).}. As phosphorylation reactions have been identified to constitute the most widespread signaling mechanism in eukaryotic cells, the set of 518 genes identified by \citeauthor{Manning2002} are a popular target for functional genomics studies. Up to 30\% of intracellular proteins may be phosphorylated, leading to 20000 phosphoprotein states, all regulated by expression of varyingly substrate specific kinases \citep{Johnson2005}. Due to the importance of kinases for cell behavior, they represent a major drug target in cancer therapeutics and might therefore also be attractive for \gls{hdt}.

In order to offset potential bias introduced by \gls{sirna} design paradigms employed by manufacturers, libraries from three different companies (Abion, Dharmacon and Qiagen) were obtained. To further account for the effect of \glspl{ote}, several \gls{sirna} sequences per target were used for both pooled and unpooled experiments. The Ambion Silencer Select Human Kinase siRNA Library V4 targets 710 genes of kinases and kinome-associated proteins with 3 sequences each, while the Qiagen HP OnGuard Human Kinase siRNA Set V4.1 comprises of 718 targets and 4 \glspl{sirna} per gene. The Dharmacon Human ON-TARGETplus siRNA Protein Kinase Libraries are designed with 715 genes in mind and are both available in 1 \gls{sirna} per well (unpooled) and 4 \glspl{sirna} per well (pooled) formats.

\begin{table}
  \centering
  \caption[Number of replicates performed for each of the pathogens and \gls{sirna} libraries.]{Number of replicates performed for each of the pathogens and \gls{sirna} libraries. The primary values indicate how many were performed in total while the value in parenthesis represents the number of screens that turned out to be unusable and had to be discarded. The effectively available number of replicates is the difference between the two. \citep{Ramo2014}}
  \label{tab:infectx-replicates}
  \footnotesize
  \begin{tabular}{L{0.2\linewidth}C{0.13\linewidth}C{0.13\linewidth}C{0.13\linewidth}C{0.15\linewidth}}
    Pathogen &
      Dharmacon (1x pooled) &
      Ambion (3x single) &
      Quiagen (4x single) &
      Dharmacon (4x single) \\
    \hline 
    \textit{B. abortus} &
      8 (1 rem.) &
      2 &
      1 &
      2 \\
    \textit{B. henselae} &
      5 (1 rem.) &
      4 (1 rem.) &
      2 (1 rem.) &
      1 \\
    \textit{L. monocytogenes} &
      4 &
      4 (2 rem.) &
      1 &
      1 \\
    \textit{S. flexneri} &
      6 (2 rem.) &
      2 &
      1 &
      1 \\
    \textit{S.} typhimurium &
      7 (4 rem.) &
      3 &
      1 &
      1 \\
    adenovirus &
      8 (1 rem.) &
      2 &
      1 &
      1 \\
    rhinovirus &
      8 (2 rem.) &
      2 &
      1 &
      1 \\
    vaccinia virus &
      2 (1 rem.) &
      2 &
      1 &
      1 \\
    \hline
  \end{tabular}
\end{table}

Depending on library and pathogen, screens were repeated 1--8 times (see table \ref{tab:infectx-replicates}). The primary values denote the total number of replicated performed and the values in parenthesis indicate the number of screens that had to be removed due to issues with transfection, weak signal intensities or usage of a protocol other than the one eventually agreed upon. The Dharmacon pooled screen was used for optimizing the assay procedures which is why for almost all pathogens some replicates had to be removed. The number of available screens is the difference between primary and parenthesized values.

\section{RNA Interference Protocols}
Central to the \gls{sirna} screens produced by InfectX was the effort to develop unified protocols for wet-tab experiments and subsequent analysis. While this approach was successfully implemented for many aspects, some deviations among the pathogens are inevitable, while others are intentionally developed to achieve similar phenotypes. Table \ref{tab:infectx-differences} summarizes some key parameters that vary between pathogens, which include seeded cell number, \gls{moi} and infection times, all optimized to yield infection rates that are of comparable magnitude. The target value for cell number was 1500 per well in order to create densely populated areas interspersed with some empty spaces, leading to cells living on colony edges and infection rates of 30--50\% were aimed for. Pathogen properties made it in some cases impossible th hit these goals and infection rate for \textit{B. abortus} remained low (\tilde 5\%), while being high in \textit{B. henselae} (\tilde 90\%) despite best efforts.

\setlength{\tabcolsep}{0.1em}
\begin{table}
  \begin{minipage}{\textwidth}
  \centering
  \caption[Differences in assay protocols among the 8 pathogens.]{Despite putting much emphasis on using identical protocols throughout all screens, some assay parameters were fine-tuned in order to obtain phenotypes such as infection and cell counts that are similar among the investigated pathogens. \citep{Ramo2014}}
  \label{tab:infectx-differences}
  \footnotesize
  \begin{tabular}{L{0.17\linewidth}C{0.09\linewidth}C{0.09\linewidth}C{0.09\linewidth}C{0.09\linewidth}C{0.09\linewidth}C{0.09\linewidth}C{0.10\linewidth}C{0.10\linewidth}}
    Pathogen &
      \mcrot{1}{l}{60}{Seeded cell number} &
      \mcrot{1}{l}{60}{\Acrshort{moi}} &
      \mcrot{1}{l}{60}{Pathogen entry time} &
      \mcrot{1}{l}{60}{Total infection time} &
      \mcrot{1}{l}{60}{DNA stain} &
      \mcrot{1}{l}{60}{Actin stain\footnote{Phalloidin-based actin stains were supplied by Dyomics and depending on absorption wavelength, different imaging channels were used: GFP for DY-495, RFP for DY-547 and Cy5 for DY-647.}} &
      \mcrot{1}{l}{60}{Pathogen stain} &
      \mcrot{1}{l}{60}{Additional stain\footnote{The Cy3 channel was used Alexa546, staining bacterially secreted InlC, while Cy5 was used for Alexa647, staining cellular \gls{il-8} during imaging.}} \\
    \hline 
    \textit{B. abortus} &
      500 &
      10000 &
      \SI{4}{\hour} &
      \SI{44}{\hour} &
      DAPI &
      DY-547 &
      GFP &
      - \\
    \textit{B. henselae} &
      300 &
      400 &
      \SI{24}{\hour} &
      \SI{24}{\hour} &
      DAPI &
      DY-547 &
      GFP &
      - \\
    \textit{L. monocytogenes} &
      600 &
      25 &
      \SI{1}{\hour} &
      \SI{5}{\hour} &
      DAPI &
      DY-647 &
      GFP &
      Alexa546 \\
    \textit{S. flexneri} &
      600 &
      15 &
      \SI{30}{\minute} &
      \SI{3.5}{\hour} &
      Hoechst &
      DY-495 &
      DsRed &
      Alexa647 \\
    \textit{S.} typhimurium &
      550 &
      80 &
      \SI{20}{\minute} &
      \SI{4}{\hour} &
      DAPI &
      DY-547 &
      GFP &
      - \\
    adenovirus &
      700 &
      0.1 &
      \SI{16}{\hour} &
      \SI{16}{\hour} &
      DAPI &
      DY-647 &
      GFP &
      - \\
    rhinovirus &
      1000 &
      8 &
      \SI{7}{\hour} &
      \SI{7}{\hour} &
      DAPI &
      DY-647 &
      GFP &
      - \\
    vaccinia virus &
      600 &
      0.125 &
      \SI{1}{\hour}\slash \SI{8}{\hour}\footnote{The two values stand for primary and secondary infection times, respectively. The same goes for pathogen dyes.} &
      \SI{24}{\hour} &
      Hoechst &
      DY-647 &
      GFP\slash RFP &
      - \\
  \end{tabular}
  \end{minipage}
\end{table}

The usage of control wells enables diagnosis of possible problems that may occur in RNAi screens, including cytotoxicity of \gls{sirna}, low transfection yields, failure of RNAi pathway induction, dominance of non-specific responses, and therefore should be embedded in every assay plate. Three types of control experiments are typically employed: positive, negative and mock (no \gls{sirna} treatment). Positive controls are used to confirm expected response, while negative controls help distinguish sequence specific from unspecific effects, and mock experiments help to establish a baseline \citep{Sittampalam2004}.

Positive controls ideally constitute of \glspl{sirna} with known effect on the assay under investigation and are therefore often unavailable beforehand. Instead, controls to check transfection efficiency and reporter quality are usually employed. One straightforward possibility for monitoring delivery is by targeting a gene that is vital to the cell. \Gls{kif11}, for example is a gene involved in cell cycle progression, the down-regulation of which is induces apoptosis. Furthermore there are mixtures of \glspl{sirna} available (e.g. AllStars Hs Cell Death Control siRNA by Qiagen) optimized for killing cells by targeting several ubiquitously expressed genes essential for cell survival. The downside of assessing transfection by killing cells is that a potentially toxic effect of the delivery mechanism itself may be masked. This can be mitigated by either performing negative control experiments (which should be done anyways) or by targeting housekeeping genes that are abundantly expressed but do not affect cell viability. Dharmacon suggests three such genes, \gls{ppib}, \gls{gapdh} and \gls{lmna}, for which they sell specially branded control \glspl{sirna} (as does Ambion).

Fluorescent dyes are also frequently employed in positive control experiments, typically by labeling \gls{sirna}, allowing for visual inspection of reagent localization within the cell (nuclear uptake indicates efficient transfection), or by targeting reporter genes. The latter method either allows for confirming that the reporting mechanism (usually expression of \gls{gfp} or luciferase) works as intended, or establishing that \gls{sirna} transfection was successful. Again, \gls{sirna} products targeting the commonly used reporter genes are readily available from manufacturers.

\renewcommand{\arraystretch}{1.1}
\setlength{\tabcolsep}{0.14em}
\begin{table}
\begin{minipage}{\textwidth}
  \centering
  \caption[Control experiments used in the different screens.]{Depending on screen type and pathogen, different genes were targeted for control experiments. Abbreviations: AU, Ambion unpooled; DP, Dharmacon pooled; DU, Dharmacon unpooled; and QU, Qiagen unpooled.}
  \label{tab:infectx-control}
  \footnotesize
  \input{R/infectx-control}
\end{minipage}
\end{table}

Negative controls should lack homology with known targets in order to separate non-specific effects from sequence specific silencing. Such \glspl{sirna} are therefore engineered to contain a passenger strand seed region (the first 2--6 nt from the 5' end) that matches no known gene and and generally have poor overall sequence identity between passenger strand and any known gene. One way of generating such a sequence is taking an assay \gls{sirna} and randomizing the order of nucleotides while keeping the nucleotide composition unchanged (i.e. scrambling the sequence). Multiple proposals are usually generated and subsequently checked for applicability by sequence alignment to the target genome. While scrambling has the advantage that a possible effect of nucleotide composition is removed, it is infeasible for large-scale screens and often a set of predefined sequences sold by manufacturers (for example Silencer Select Negative Control from Ambion, ON-TARGETplus Non-targeting Control siRNAs from Dharamcon or AllStars Negative Control siRNA from Qiagen) is used instead (while still being called scrambled controls). 

Table \ref{tab:infectx-control} lists the \gls{sirna} control experiments that were performed throughout all screens and indicates the set of controls each screen contains. Common to all pathogens and all libraries are the previously mentioned positive control Kif11, one of two \gls{gfp} targeting sequences, scrambled \glspl{sirna}, as well as mock experiments. Additionally, several controls for general infection mechanisms were included in most screens, including ATP6V1A (a \ce{H^+} transporting ATPase, responsible for acidification of endocytic and lysosomal vesicles), ARPC3 (\acrshort{arp23}), and \gls{cdc-42}, both part of actin-dependent processes surrounding pathogen uptake and \gls{abm}. Some controls however are specific to a subgroup of pathogens or single pathogens and include the following (grouped by pathogen).
\begin{description}[leftmargin=0.5cm]
\item[\textit{Bartonella}\slash \textit{Brucella}:] The small GTPase Rab2 is required for anterograde \gls{er}--Golgi transport, capable of interacting with RicA of \textit{B. abortus} \citep{DeBarsy2011} and TLN1 (talin 1) has been determined to be necessary for invasome formation in \textit{B. henselae} infection via \textbeta\textsubscript{1} signaling \citep{Truttmann2011}.
\item[\textit{Listeria}:] During endocytosis of \textit{L. monocytogenes}, the ubiqutin ligase Cbl is a recruited to the site of entry and seems to be involved in InlB mediated, clathrin dependent (CLTC encodes the heavy chain 1 component of clathrin) bacterial uptake \citep{Veiga2005}. DNM2 (dynamin-2) is a further protein involved in host entry and PIK3R1 is a phosphatidylinositol 3-kinase, downstream to Met, the cellular receptor to InlB.
\item[\textit{Salmonella}:] The actin-modulating protein CFL1 (cofilin-1), responsible for actin depolymerization, the cellular integrin receptor ITGAV and RAB7A, a small GTPase that regulates vesicular trafficking, have all been shown to be hits in a salmonella invasion screen \citep{Misselwitz2011}.
\item[\textit{Shigella}:] Down regulation of ARF1 (a \gls{gtp} binding protein involved in vesicle trafficking) and phosphatidylinositol 4-kinase PI4KA has been suspected of interfering with pathogen entry, while suppression of CHUK\slash NOD1 (both involved in \gls{nf-kb} signaling) may inhibit \gls{il-8} production by uninfected bystander cells, thereby possibly promoting infection \citep{Kasper2012}. 
\item[Adenovirus\slash rhinovirus:] SNX9 regulates dynamin assembly and is therefore crucial to viral endocytosis and PRKCA is a Ser-\slash Thr-specific protein kinase responsible for a wide array of regulatory signals. PRKCA might be involved in influenza virion buddying and has been implicated in playing a role in intracellular proliferation of hepatitis \citep{Kanehisa2000}.
\item[Vaccinia virus:] The Ser-\slash Thr-specific protein kinase PRKCA is targeted by \gls{cdc-42}\slash \gls{rac-1} and has been shown to be required for \gls{mv} entry \citep{Mercer2008}. Up-regulation of PSMA6 (a proteasome complex component) enables evasion of intracellular immune surveillance \citep{Zhou2014} and TSG101, ubiquitin-conjugating enzyme, hampers \gls{ev} production \citep{Honeychurch2007}. 
\end{description}

Many of these pathogen specific positive control targets are not well established and while some have been previously been identified and validated, others represent best guesses and might not serve their purpose particularly well. Controls wells are typically located at the plate border (rows A and P; columns 1, 2, 23 and 24) and the different control experiments were replicated multiple times on each plate.

For all screens, \gls{sirna} transfection was carried out by adding \SI{25}{\micro\litre} of RNAi\-MAX\slash DMEM (\SI{0.1}{\micro\litre}\slash \SI{24.9}{\micro\litre}) transfection agent to \SI{1.6}{\pico\mol} siRNA diluted in \SI{5}{\micro\litre} RNase-free \ce{ddH2O} contained in each of the 384 wells per screening plate. After \SI{1}{\hour} incubation at room temperature, the required number of cells were added (see table \ref{tab:infectx-differences}), suspended in \SI{50}{\micro\litre} DMEM\slash 16\% \gls{fbs}. Plates were subsequently incubated for \SI{72}{\hour} at \SI{37}{\celsius} and 5\% \ce{CO2}, followed by the pathogen specific infection procedure (see section \ref{sec:pathogen-protocols} for details). Following infection, cells were fixed in \gls{pfa} and stained for DNA, F-actin and additional pathogen specific markers (see table \ref{tab:infectx-differences}). Plates were sealed prior to imaging.

\section{Image Acquisition and Data Processing}
Imaging was performed both at the University of Basel and the Light Microscopy and Screening Center of ETH Z\"urich, using ImageXpress micro (IXM) HCS wide-field microscopes from Molecular Devices, equipped with Thermo Scientific CataLyst Express robotic plate handlers capable of storing and serving up to 45 plates. Lumencor Spectra X solid-state light engines (LED light sources), 10x S Fluor objective lenses by Nikon with a numerical aperture of 0.45 and Photometrics CoolSNAP HQ 14-bit CCD cameras, resolving $1392 \times 1040$ pixels (individual pixel size of $\SI{6.45}{\micro\meter} \times \SI{6.45}{\micro\meter}$), complete the hardware setup. Channel selection is assay specific and stain dependent Semrock filters (DAPI\slash Hoechst, GFP\slash FITC\slash Alexa488, Cy3, Cy5, Quadband DAPI-GFP-mCherry-Cy5) are employed (see table \ref{tab:infectx-differences} for details).

Wells are divided into $3 \times 3$ grids for most plates while some $2 \times 3$ site images exist too, with no spacing and no overlap, and Molecular Devices MetaXpress High-Content Image Acquisition and Analysis Software was used for recording images. Software parameters include no gain, well bottom as autofocus target, site-specific autofocusing, enabling of laser-based focusing and no shading correction. For each imaging channel, focus Z-offset was selected manually and exposure time was automatically calculated. In cases of poor dynamic range or overexposure, manual correction to exposure time was applied. Upon imaging, data was transferred to iBrain2\slash screeningBee \citep{Rouilly2012} for further processing.

\subsection{Data Handling (iBrain2\slash screeningBee)}
In case of microscopy based \gls{sirna} screening experiments, a complex task of data handling and processing follows the imaging stage. A wealth of data is generated by imaging devices\footnote{A genome-wide screen (\tilde 27000 individual experiments) in 384 well format involves 70--100 plates depending on the number of controls per plate. Each plate yields 10000--14000 images (384 wells, 9 sites and 3--4 channels), leading to 700000--1400000 individual images and requiring multiple terabytes of storage.} which has to be accessibly and redundantly stored. Furthermore, analysis of image data is a processing intensive task that quickly becomes reliant on \gls{hpc} resources, entailing specialized requirements due to the oftentimes shared and centralized nature of such systems. The authors of iBrain2 summarize the key steps in RNAi \gls{hcs} data processing as follows \citep{Rouilly2012}:
\begin{enumerate}
\item \textbf{Data acquisition.} The raw data of \gls{sirna} screens is produced as digital images by microscopy. Acqusition times of several hours per plate are typical and a single plate yields \tilde \SI{20}{\giga\byte} of data.
\item \textbf{Permanent storage.} Due to infeasibility of re-screening plates, all raw data has to be stored in a sufficiently redundant manner. Furthermore, the permanent data store has to be able to serve portions of the dataset quickly and efficiently.
\item \textbf{Temporary data staging to \gls{hpc}.} In order for the compute cluster not to be bound by network latency, it is often necessary to stage the data to be analyzed to local scratch space. This step becomes superfluous whenever the permanent storage system is directly integrated in the cluster's high-speed network.
\item \textbf{Data analysis.} Many aspects of processing a large number of images are embarrassingly parallel and a cluster environment is ideally suited to tackle this computation intensive task.
\item \textbf{Permanent storage of results.} Some results produced by the analysis procedure will be saved back to the permanent storage system. While it may be sensible to save storage space and carry out some procedures on the fly, this is not feasible for all analysis routines.
\item \textbf{Publication and archiving.} Upon completion of the project, some data will be made availably publicly and all data worth keeping is moved to an archival system capable of cheap long-term storage where quick retrieval is unimportant.
\end{enumerate}

The requirements outlined above led to the development of iBrain2 as a modular workflow manager capable of setting up reproducible procedures. The software is implemented in Java and uses an XML-based format for defining workflows. Furthermore, it interfaces with other open source projects, such as openBIS \citep{Bauch2011} which can be used as storage system and CellProfiler \citep{Carpenter2006,Kamentsky2011}, a popular tool for analyzing cell based image material. Within InfectX, not iBrain2 itself is used, but a derivative thereof called screeningBee, featuring tighter openBIS integration and a large set of CellProfiler extensions.

\subsection{Image Analysis (CellProfiler)}
Prior to the development and release of CellProfiler in 2006, large-scale screens were routinely evaluated though tedious and labor intensive visual inspection by expert biologists. While humans are able to perceive and contextualize image data in ways that computers still struggle to emulate, a lot of detail is lost though human interpretation. Small differences are hard to spot, leading to subtle patterns being missed and humans focus on a handful of features, not being able to take into account the plethora of information encoded in crowded images. Furthermore, human study of image data leads to qualitative evaluations, while computational analysis yields a potentially large number of quantitative measures. Lastly and perhaps most importantly, human based assessment of imagery coming from a project the size of InfectX is simply infeasible. In the words of \citeauthor{Carpenter2006}, the authors of CellProfiler, consequences of implementing computational image analysis for \gls{hts} are as follows:

\begin{quote}
With the successful application of sophisticated image analysis methods, the bottleneck of image-based genome-wide screens is now moving downstream to data visualization, exploration, and statistical analysis in order to accommodate the number and richness of measurements that result from image-based genome-wide assays.
\end{quote}

CellProfiler is an open-source software solution catering to the needs of \gls{hts} screening initiatives and providing a customizable, modular platform for cellular image analysis. CellProfiler processing is based on modules which are placed in sequential to form a pipeline, usually starting out with image processing, followed by object identification and concluded by calculation of feature measurements on these objects. The pipeline's modules and settings can be saved to a configuration file, in order to ensure reproducibility and transferability of workflows. The original version of CellProfiler was written in MATLAB and a more recent version 2 has been ported to python. InfectX procedures are based on version 1 \citep{Carpenter2006}.

The InfectX image analysis routine starts out with metadata collection and image quality assessment modules, followed by shading correction. Detecting out of focus images, wells that show signs of experimental error or problematic plates, however is still largely carried out by humans. Position dependent illumination and sensor inhomogeneities need to be corrected for, as they lead to decreased sensitivity in darker areas and make comparison of intensity based features between objects belonging to different areas of the image problematic.

Shading correction, while offered by a default module of CellProfiler, is implemented in a module developed specifically for InfectX. A shading model is generated per microscope, lamp, channel, pathogen and assay which can simply be applied to all corresponding images. This reduces the computational complexity of shading correction significantly and avoids errors associated with an automated procedure being applied to a diverse range of instances. Shading models are produced by overlaying images and calculating histograms per pixel position, yielding robust estimates for foreground intensities and by fixed offsetting, also estimates for background intensities. Subsequently, a foreground model is calculated and applied, leading to conservatively corrected images. With only little shading remaining, reliable object segmentation is possible, the result of which are used to calculate a further, final shading model based on pixel histograms and average object intensities.

Illumination corrected images are stitched together, channel separated and forwarded to CellProfiler modules for segmentation and feature calculation. Cell identification is a hard problem, especially if clumps are present. Whenever objects are present that can easily be identified, such as nuclei if DNA was stained, these are identified first (primary objects), facilitating identification of secondary objects, e.g. cells. In case of clumped cells, a three step algorithm is employed: clumps are recognized, then the dividing lines are defined and finally, some resulting objects may be removed or merged based on measurements such as size or shape.

Several modules developed in-house are used during the feature extraction stage, extending CellProfiler functionality to specific requirements including invasome and bacterial aggregate detection, more efficient neighbor feature measurement and collection of various properties per sub-cellular object. The features calculated by the CellProfiler pipeline are explored more in-depth in section \ref{sec:scf-data}. All results produced by image analysis are saved to an openBIS instance.

\begin{figure}
  \centering
  \includegraphics[width=0.7\textwidth]{openbis}
  \caption[Overview of the openBIS architecture and a more in-depth look at the data storage implementation.]{An openBIS instance is deployed as a service consisting of an application server (AS) and one or more data store servers (DSSs). At both levels, clients can interact with the system to query, fetch or deposit data. \citep{Bauch2011}.}
  \label{fig:openbis}
\end{figure}

\subsection{User Accessible Data Storage (OpenBIS)}
The authors of openBIS (open Biology Information System) argue that the availability of a domain specific information management system plays an important role in data-driven biological research, serving as a basis for management of large amounts of experimental data and acting as both a source and a sink for analysis procedures. Furthermore, such a system should provide a user interface capable of visualizing complex dependency structures and a set of \glspl{api} for accessing the stored data programmatically \citep{Bauch2011}. Finally, it should be possible to make a subset of the data available to the public, providing a simplified interface of the browsing and searching capabilities.

In order for openBIS to be applicable to the broad range of scenarios it is intended for, extensibility and transparent interfaces are paramount, as is scalability and performance. To address the latter aspect, separation of bulk data from metadata is a guiding principle. Both ingestion of high-throughput, high-content data as well as serving it to users and analysis routines is CPU, I/O and network intensive and therefore should not be performed on the same system as user-facing metadata queries which only deals with comparably inexpensive tasks that scale well. Additionally, scalability requires that bulk data storage may be distributed among several systems.

With these requirements in mind, openBIS is designed to consist of an \gls{as}, dealing with metadata, serving user queries such as searches and visualizing dependencies via the browser based graphical user interface, as well as one or several \glspl{dss} that handle the bulk data. Communication between the two layers is asynchronous.

All datasets in openBIS are immutable. While this is potentially wasteful in terms is storage resources, it can be argued that storage space is cheap and such an architecture improves traceability of constant evolution of analysis procedures and corresponding results. Datasets are subject to a hierarchical structure paradigm consisting of the entities \begin{enumerate*}[label=\itshape (\arabic*)] \item\textit{Data Space}, \item\textit{Project}, \item\textit{Experiment}, and \item\textit{Sample} \end{enumerate*}. In the example of \gls{hts} data, a sample represents a plate and associated datasets include raw images, feature measurements and infection scores. Metadata can be provided either in unstructured form which is handled as attachments associated with project, experiment or sample entities, or as structured objects (consisting of name, label, description and value) that are linked to experiments, samples and datasets.

Data storage by a \gls{dss} follows a hybrid model consisting of a relational database, storing index information, file metadata and selected results, in addition to flat-file data store which can be distributed between several file systems. Management of fill state of the associated storage shares and data distribution is all managed by the individual \glspl{dss}. Some data presentation task handled by \glspl{dss} include data visualizations such as plate heatmaps and assembly of multichannel images, as well as overlay of analysis results (e.g. segmentation).

Ingestion of new data can either be triggered by web import, custom software talking to the corresponding \gls{api} or simply by adding to folders that are monitored by a \gls{dss} as dropboxes. Responding to write activity, the highly customizable ETL (extract, transform, load) routine extracts metadata, creates datasets annotated with metadata in the \gls{as} database, links them to the appropriate entities and adds the new datasets to the data store. Export of data is either browser based or can be handled by provided command line tools, accessing openBIS via a Java service facade. Two separately designed tools are integrated for used with openBIS: CIFEX and Datamover. The former enables browser based transfer of large files and supports interruption  of transfers as well as checksumming for file integrity, while the latter can be used to automatically transfer data off of an instrument computer into a dropbox folder via SSH or rsync protocols.

\section{Single Cell Feature Data}
\label{sec:scf-data}
The set of accessible single cell features varies among screens and pathogens to some extent. While there are some general features, such as geometry, intensity and texture of nuclear and cellular objects available for every plate dataset, others, including invasome and \gls{il-8} features are specific to a subset of the pathogens, while others still, such as certain neighbor features or Vononoi cell segmentation were only recently added to the feature extraction pipeline and therefore have not yet been calculated for all screens. Such discrepancies are not treated by this overview which focuses on general aspects of currently available single cell features.

Nomenclature of single cell features follows a hierarchical pattern consisting of an object specifier, a measurement group, the measurement name itself and the imaging channel it was applied to, as follows:
\begin{center}
\texttt{<object>.<group>\_<measurement>\_<channel>}
\texttt{Cells.Intensity\_MedianUpperTenPercentIntensity\_CorrDNA}
\end{center}
In the above example, the measured objects are cells, the feature is of intensity type, the feature name indicates that it represents the mean of intensities above the upper decile and the channel description reveals that DNA stain values are used. Throughout the different screens, 14 single cell feature objects are available:

\begin{figure}
  \centering
  \includegraphics[width=0.95\textwidth]{segmentation}
  \caption[Object detection of nuclei, perinuclei, voronoi cells and cell bodies along with potential pitfalls.]{Object identification is started with easily recognizable nuclei, which become primary objects and serving as basis for detecting secondary objects. Problems with nuclei detection include superfluous splitting (a) and failure to separate some touching objects (c). Mostly results are acceptable and some touching nuclei are correctly separated (b). Perinuclei extend the nuclear border and therefore suffer from problems introduced in nuclear recognition. Further issues include overlap with extracellular space (d) and small perinuclear area in densely populated areas (e). Voronoi cells further extend the nuclear area and often grow beyond cell boundaries (f). Finally, cell segmentation is a hard problem due to irregular shapes, variability in size and inhomogeneities in actin distribution (h). Results suffer from falsely identified nuclei (g and i).}
  \label{fig:segmentation}
\end{figure}

\begin{description}[leftmargin=0.5cm]
\item[Nuclei:] Using the DNA channel, nucleus recognition is fairly accurate owing to generally uniform morphology of nuclei, uniform distribution of DNA and good contrast of DNA staining. The current implementation for segmentation is based on Otsu's method which converts a monochromatic image into a binary image by finding an appropriate threshold, effectively separating foreground from background. Objects that have an Euler characteristic $\chi \le 0$ are filled and apparently clumped objects are split. Finally, objects outside of an allowed size range are discarded (mostly dust particles which are too big and pathogen DNA which yields objects that are too small). Nuclei represent primary objects with serve as basis for the detection of secondary objects (see figure \ref{fig:segmentation}, A).
\item[PeriNuclei:] The perinuclear region is defined as an extension of the nuclear border by 8 pixels and removal of the nucleus itself. As some pathogens establish their replicatory niche in this region it lends itself to closer investigation. Furthermore the border is only dependent on the reliably established nuclear object. Potential pitfalls include extension of the perinuclear region into extracellular space or neighboring cell when the nucleus is located close to the plasma membrane, as well as reduced area whenever cells are clustered, as perinuclear regions may not overlap (see figure \ref{fig:segmentation}, B).
\item[VoronoiCells:] Simultaneously extending all nuclei by maximally 24 pixels or until a neighboring border is met leads to Voronoi cells. This type of defining cell bodies is very prone to extending into extracellular space, as the actin channel is not taken into account (see figure \ref{fig:segmentation}, C).
\item[Cells:] The cell body border is determined by propagation segmentation based on thresholding. Starting from the nucleus, the cytoplasmic region is extended until the intensity gradient changes abruptly, indicating either the beginning of background, or the boundary to a neighboring cell. Non-uniform actin intensity patterns, irregular shape, large variability in size and frequent touching of neighbors makes this a hard problem and results can be unreliable. Common to all secondary objects, errors in determining primary objects will propagate and lead to cells erroneously being cut up or aggregated. Furthermore, borders to neighbors are frequently not determined correctly, but significant discrepancies among human expert labeling are to be expected in such situations as well (see figure \ref{fig:segmentation}, D).
\item[ExpandedNuclei:] Serving as precursors to PeriNuclei objects, these objects are not further quantified and only have coordinates and number of perinuclear children (almost exclusively 1) associated.
\item[Bacteria/Viruses/Pathogen:] Due to the heterogeneity between pathogens, detection is assay specific. Generally, pathogens are primary objects and not derived of cell nuclei. In \textit{Bartonella} and \textit{Brucella} screens, large and small pathogen clusters are segmented using Otsu's method on the pathogen channel. Individual bacteria in \textit{Listeria} screens, \glspl{scv} in \textit{Salmonella} screens and small\slash medium pathogen clusters in both \textit{Shigella} and rhinovirus screens are identified using wavelet decomposition of the pathogen channel, while for adenovirus screens, detection of large pathogen aggregates relies on propagation outwards from nuclei, much like cell body identification but using the pathogen channel. For vaccinia virus, direct pathogen segmentation is currently not possible.
\item[IntBacteria/ExtBacteria:] In \textit{Bartonella} assays, different stains were used for intracellular (\acrshort{gfp}) and extracellular (Cy5) bacteria which consequently can be distinguished during image analysis. This is helpful to prevent bacteria the are sit on top or below a cell from being considered intracellularly located.
\item[BlobBacteria:] Specific to \textit{Brucella} screens, these large bacterial aggregates do not have any real measurements associated, except coordinates.

\begin{figure}
  \centering
  \includegraphics[width=0.95\textwidth]{actin-features}
  \caption[Detection of two actin based structures, CometTails and Invasomes.]{Both CometTails and Invasomes are detected on actin channels, which poses difficulties due to noisy background. Invasomes are characterized by their ring-shaped morphology (a), while CometTails appear as actin streaks (c). Currently invasome detection yields many false-positives (b).}
  \label{fig:actin-features}
\end{figure}

\item[Invasomes:] As one of two internalization structures in \textit{Bartonella} infection, invasomes are characterized as actin surrounded large bacterial aggregates. Exploiting this morphology, the current invasome detection uses a ring template to search the actin channel for candidates and exploits the intensity difference to the cell mean for confirmation. Unfortunately, this yields many false positives. Improvements are being discussed but have yet to be implemented (see figure \ref{fig:actin-features}, A).
\item[IL8:] In \textit{Shigella} assays, cellular \gls{il-8} was stained in order to study proinflammatory response of uninfected bystander cells. Due to these objects having their own channel, segmentation is fairly straightforward.
\item[CometTails:] Specific to \textit{Listeria} screens, the telltale signs of \gls{abm} can be identified as dense strokes on the actin channel. CometTails contain only few measurements beyond their location (see figure \ref{fig:actin-features}, B).
\item[Neighbors:] While not representing detected objects themselves, features involving identities of adjacent objects are summarized within this category. Currently, indices of Cells and Bacteria that come closer than 2 pixels and of Voronoi cells that do not lie further apart than 8 pixels are collected. These thresholds compensate for slight segmentation errors. Furthermore, for each pair of detected neighbors, the length of common border is recorded. Taken together, features of this category allow construction of both a binary adjacency matrix and a weighted adjacency matrix corresponding to each of the objects that were considered.
\end{description}

Once images are segmented into objects, measurements are applied to the individual image areas. Dependent on object class, the number and types of available features differ, sometimes owing to systematic causes and at other times, only due to differences in implementations of per-object feature recognition. The most commonly available features are grouped into categories \textit{AreaShape}, \textit{Intensity}, \textit{Texture}, \textit{RadialDistribution}, \textit{Location}, \textit{Neighbors\slash IdentityOfNeighbors\slash Per\-centTouchingNeighbors} and \textit{Parent\slash Children}.

\renewcommand{\arraystretch}{1.5}
\setlength{\tabcolsep}{0.2em}
\begin{table}
  \begin{minipage}{\textwidth}
  \centering
  \caption[List of AreaShape features with corresponding descriptions.]{List of AreaShape features with corresponding descriptions. Some information is taken from the CellProfiler manual \citep{Carpenter2006}.}
  \label{tab:areashape-feats}
  \footnotesize
  \begin{tabular}{L{0.25\linewidth}L{0.7\linewidth}}
    Feature name &
      Feature description \\
    \hline 
    Area &
      The number of pixels enclosed by the object border. \\
    Eccentricity\footnote{\label{fn:ellipse} The corresponding ellipse is defined as the best fitting ellipse in the sense that it has identical second moments compared to the original object \citep{Rocha2002}.} &
      Ratio of the distance between the foci and major axis length of the corresponding ellipse. The value is between 0 (circle) and 1 (line segment).\\
    EulerNumber &
      1 minus the number of holes within the object, assuming 8-connectivity. \\
    Extent &
      Area of the object divided by the area of the bounding box. \\
    FormFactor &
      Calculated as $4\pi*area/perimeter^2$. Equals 1 for a perfectly circular object. \\
    <Spec>AxisLength\textsuperscript{\ref{fn:ellipse}} &
      Spec\,=\,\{Major, Minor\}; The major\slash minor axis length (in pixels) of the corresponding ellipse. \\
    Orientation\textsuperscript{\ref{fn:ellipse}} &
      The angle (ranging from \SI{-90}{\degree} to \SI{90} {\degree}) between the x-axis and the major axis of the corresponding ellipse \\
    Perimeter &
      The total number of pixels around the object boundary. \\
    Solidity &
      Proportion of the pixels in the convex hull that are also part of the object, i.e. $area_{object}/area_{conv. hull}$. \\
    <Sub\footnote{\label{fn:subcellular} Naming of these features currently is fairly heterogeneous with possible prefixes \textit{PerObj}, \textit{SubCellBacteria}, \textit{SubCellViruses}, \textit{SubCell}. All are based on the set of subcellularly located pathogens.}>Area &
      The total number of pixels enclosed by all subcellular pathogen object borders. \\
    <Sub\textsuperscript{\ref{fn:subcellular}}><Stat\footnote{\label{fn:statistics} The statistics applied in order to describe the underlying distributions are mean, median, lower and upper quartiles, standard deviation and sum.}>Distance\-ToNuclNorm &
      Distances between the nucleus and all subcellular pathogen objects are normalized and a summary statistic is applied. \\
    <Sub\textsuperscript{\ref{fn:subcellular}}><Stat\textsuperscript{\ref{fn:statistics}}>Distance\-ToNucl &
      Distances from nucleus to all subcellular pathogen objects are calculated and a summary statistic is applied.\\
    <Sub\textsuperscript{\ref{fn:subcellular}}><Stat\textsuperscript{\ref{fn:statistics}}>Pairwise\-Distance &
      A summary statistic is applied to all pairwise distances between subcellular pathogen objects. \\
    <Sub\textsuperscript{\ref{fn:subcellular}}><Stat\textsuperscript{\ref{fn:statistics}}>Shortest\-Distance &
      For each subcellular pathogen object, the shortest distance among all distances to other subcellular objects is selected and a summary statistic is applied to the collection of shortest distances. \\
  \end{tabular}
  \end{minipage}
\end{table}

Features belonging to the AreaShape category are not applied to a specific imaging channel but serve to summarize geometric properties of detected objects (\textit{cf.} table \ref{tab:areashape-feats}). When applied to actin channel objects that cannot reliably be segmented, such as cell bodies, some caution has to be exerted, as these features are strongly affected by upstream errors. Nevertheless, this group constitutes features that are among the most frequently used for analysis procedures. Some infection phenotypes, for example membrane ruffling encountered during trigger-type host entry mechanisms are captured by some of these features (FormFactor is able to measure raggedness).

\begin{figure}
\centering
  \begin{tikzpicture}
    \GraphInit[vstyle=Classic]
    \SetVertexNoLabel
    \tikzset{VertexStyle/.style = {shape = circle, fill = black, minimum size = 5pt, inner sep=0pt}}
    \grEmptyPath[form=2,x=0,y=0,RA=0.5,rotation=90]{11}
    \grEmptyPath[form=2,x=3,y=0,RA=1,rotation=90,prefix=b]{6}
    \SetUpEdge[lw=0.75pt,color=black]
    \Edge(a10)(b0)
    \Edge(a9)(b0)
    \Edge(a8)(b0)
    \Edge(a7)(b0)
    \Edge(a6)(b0)
    \Edge(a5)(b0)
    \Edge(a4)(b0)
    \Edge(a3)(b0)
    \Edge(a6)(b1)
    \Edge(a7)(b1)
    \Edge(a6)(b2)
    \Edge(a7)(b2)
    \Edge(a6)(b3)
    \Edge(a7)(b3)
    \Edge(a10)(b4)
    \Edge(a9)(b4)
    \Edge(a8)(b4)
    \Edge(a7)(b4)
    \Edge(a5)(b4)
    \Edge(a10)(b5)
    \Edge(a9)(b5)
    \Edge(a8)(b5)
    \Edge(a7)(b5)
    \Edge(a5)(b5)
    \tikzset{AssignStyle/.append style = {left=6pt}}
    \AssignVertexLabel[size=\footnotesize]{a}{UpperTwoPercent,UpperFivePercent,UpperTenPercent,LowerQuartile,UpperQuartile,Std,Median,Mean,Min,Max,Integrated}
    \tikzset{AssignStyle/.append style = {right=12pt}}
    \AssignVertexLabel[size=\footnotesize]{b}{\textless Sub\textgreater\ (e.g. SubCellLowerQuartile),UpperTwoPercent (e.g. MeanUpperTwoPercent),UpperFivePercent (e.g. MeanUpperFivePercent),UpperTenPercent (e.g. MedianUpperTenPercent),Edge (e.g. MaxIntensityEdge),Weighted (e.g. IntegratedWeighted)}
  \end{tikzpicture}
  \caption[Intensity features and possible modifiers.]{Common summary statistics that are applied to intensity data, along with their possible modifiers. The subcellular intensity features have different prefixes, including SubCellBacteria, SubCellCometTails, SubCellViruses, PerObject and SubCell.}
  \label{fig:intensity-features}
\end{figure}

Intensity features mostly consist of summary statistics being applied to all or a portion of per channel intensity data within a given object. The functions used to describe the underlying distributions are sum (Integrated), lower quartile, maximum (Max), mean, median, minimum (Min), standard deviation (Std), upper quartile, upper vigintile (UpperFivePercent), upper decile (UpperTenPercent) and upper 50-quantile (UpperTwoPercent). Additional features are generated by using certain functions only on the object border pixels, yielding IntensityEdge features (e.g. IntegratedIntensityEdge) and by only including the upper 10\%, 5\% and 2\% of the data. Weighted features are calculated by extracting weights from a corresponding grayscale image and multiplying individual intensity values before applying the summary statistic. Restricting the data to subcellular pathogen objects yields features such as SubCellLowerQuartile. For a complete list of modifiers and implemented combinations, refer to figure \ref{fig:intensity-features}. The only currently available intensity feature that is not shown in \ref{fig:intensity-features} is MassDisplacement, which quantifies the distance between the centers of gravity in the gray-level representation of the object and the binary representation of the object.

All texture measurements are either based on Haralick features \citep{Haralick1973} or Gabor wavelet features \citep{Gabor1946}, as implemented in CellProfiler and are calculated per image channel. The gray-level co-occurrence matrix $G_k$ with dimensions $n \times n$ where n is the number of gray levels serves as basis for Haralick's texture features \citep{Carpenter2006}.

\begin{equation}
  G_k = \begin{bmatrix}
    p(1,1) & p(1,2) & \cdots & p(1,n) \\
    p(2,1) & p(2,2) & \cdots & p(2,n) \\
    \vdots  & \vdots  & \ddots & \vdots  \\
    p(n,1) & p(n,2) & \cdots & p(n,n) 
  \end{bmatrix}
\end{equation}

Element $p(i,j)$ is calculated by counting the number of times, a pixel with value $i$ is adjacent to a pixel with value $j$, divided by the total number of comparisons. Consequently, $p(i,j)$ can be thought of as representing a probability. Moreover, assuming 8-connectivity, 4 different types of adjacency are possible: horizontal, vertical, top left to bottom right diagonal and top right to bottom left antidiagonal, yielding 4 co-occurrence matrices $G_k, k\in\{1,2,3,4\}$. 14 statistics were developed by \citeauthor{Haralick1973}, 13 of which are included with CellProfiler:

\begin{align*}
p_x(i) &= \sum_{j=1}^{n} p(i,j), & p_y(j) &= \sum_{i=1}^{n} p(i,j), \\
p_{x+y}(k) &= \mathop{\sum_{i=1}^{n}\sum_{j=1}^{n}}_{i+j=k} p(i,j), &
  p_{x-y}(k) &= \mathop{\sum_{i=1}^{n}\sum_{j=1}^{n}}_{\lvert i-j\rvert=k} p(i,j), \\
  \mu_x &= \Erw[p_x], & \mu_y &= \Erw[p_y], \\
  \sigma_x &= \sqrt{\Var(p_x)}, & \sigma_y &= \sqrt{\Var(p_y)}, \\
  HX &= -\sum_{i=1}^{n} p_x(i) \log\left(p_x(i)\right), &
  HY &= -\sum_{i=1}^{n} p_y(i) \log\left(p_y(i)\right),
\end{align*}
\begin{align*}
  HXY1 &= -\sum_{i=1}^{n} \sum_{j=1}^{n} p(i,j) \log\left(p_x(i)p_y(j)\right), \\
  HXY2 &= -\sum_{i=1}^{n} \sum_{j=1}^{n} p_x(i)p_y(j) \log\left(p_x(i)p_y(j)\right)
\end{align*}
\begin{align}
\text{AngularSecondMoment} & &
  f_1 &= \sum_{i=1}^n \sum_{j=1}^n p(i,j)^2 \\
\text{Contrast} & &
  f_2 &= \sum_{k=0}^{n-1} k^2 \left(\mathop{\sum_{i=1}^{n}\sum_{j=1}^{n}}_{\lvert i-j\rvert=k} p(i,j)\right) \\
\text{Correlation} & &
  f_3 &= \frac{\sum_{i=1}^n \sum_{j=1}^n (ij) p(i,j)-\mu_x\mu_y}{\sigma_x\sigma_y}\\
\text{Variance} & &
  f_4 &= \sum_{i=1}^n \sum_{j=1}^n (i-\mu)^2 p(i,j)\\
\text{InverseDifferenceMoment} & &
  f_5 &= \sum_{i=1}^n \sum_{j=1}^n \frac{1}{1+(i-j)^2} p(i,j)\\
\text{SumAverage} & &
  f_6 &= \sum_{i=2}^{2n} i p_{x+y}(i)\\
\text{SumVariance} & &
  f_7 &= \sum_{i=2}^{2n} (i-f_8)^2 p_{x+y}(i)\\
\text{SumEntropy} & &
  f_8 &= -\sum_{i=2}^{2n} p_{x+y}(i) \log\left(p_{x+y}(i)\right)\\
\text{Entropy} & &
  f_9 &= -\sum_{i=1}^n \sum_{j=1}^n p(i,j) \log\left(p(i,j)\right)\\
\text{DifferenceVariance} & &
  f_{10} &= \sum_{i=0}^{n-1} i^2 p_{x-y}(i)\\
\text{DifferenceEntropy} & &
  f_{11} &= -\sum_{i=0}^{n-1} p_{x-y}(i) \log\left(p_{x-y}(i)\right)\\
\text{InfoMeas1} & &
  f_{12} &= \frac{f_9-HXY1}{\max\{HX,HY\}}\\
\text{InfoMeas2} & &
  f_{13} &= \left(1-\exp\left[-2(HXY2-f_9)\right]\right)^{\frac{1}{2}}
\end{align}

The individual statistics are averaged over the four co-occurrence matrices to yield a single value per object per imaging channel. Computation of Haralick features is comparably inexpensive and the resulting texture characterization has successfully been used in image-based classification scenarios. Gabor filters on the other hand are more complicated and their application can incur signinificant computational cost. They quantify striped texture that is parallel to a given angle and sets of Gabor filters at certain angles are frequently used for pattern recognition tasks such as optical character recognition or fingerprint recognition. The directions for which Gabor filters are applied to segmented objects are the x- and y-axes, yielding the features GaborX and GaborY.

RadialDistribution features contain 3 measurements that are applied to each image channel. Given an object, it's area is divided into bins formed by concentric rings, where the center is the point that is the farthest from any edge. FracAtD records the fraction of total stain contained in a given bin, MeanFrac normalizes the fraction of total stain with the fraction of pixels of a given bin and RadialCV determines the coefficient of variation of intensity within a ring, calculated over 8 slices. Features of this category are calculated for invasome objects (4 bins) in \textit{Bartonella} screens and are used for infection scoring. Invasomes are characterized by a ring like appearance and the concentration of intensity within this ring can be detected by the FracAtD feature.

Location features store the coordinates of the center for each object (Center_X, Center_Y), while Parent\slash Children features 

\textit{Location}, \textit{Neighbors\slash IdentityOfNeighbors\slash Per\-centTouchingNeighbors} and \textit{Parent\slash Children}.

\section{Infection Scoring}
\label{sec:infection-scoring}

\begin{figure}
\centering
\begin{tikzpicture}

\node [decision={$> 0.075$}{$\le 0.075$}] {
  \texttt{Cells.Intensity\_MeanIntensity\_\\CorrPathogen}
}
[decision tree]
child { node [outcome] { infected } }
child { node [decision={$> 0.085$}{$\le 0.085$}] { 
\texttt{PeriNuclei.Intensity\_MeanIntensity\_\\CorrPathogen} } 
  child { node [outcome] { infected } }
  child { node [decision={$> 0.115$}{$\le 0.115$}] { 
  \texttt{Nuclei.Intensity\_MeanIntensity\_\\CorrPathogen} } 
    child { node [outcome] { infected } }
    child { node [outcome] { not infected } }
  }
};

\end{tikzpicture}
\caption[Decision tree for adenovirus infection scoring.]{For adenovirus infection scoring, the decision tree classifier checks if enough pathogen is detected within the cell body, the perinuclear region or the nucleus. The threshold decreases as the region of interest concentrates on areas associated with progressively involved in infection.}
\end{figure}

\begin{figure}
\centering
\begin{tikzpicture}

\node [decision={$> 0.800$}{$\le 0.800$}] { 
  \texttt{Invasomes.Intensity\_MaxIntensity\_\\Corr1Pathogen}
}
[decision tree]
child { node [outcome] { infected } }
child { node [decision={$> 0.075$}{$\le 0.075$}] { 
\texttt{Invasomes.Intensity\_MeanIntensity\_\\Corr1Pathogen} } 
  child { node [outcome] { infected } }
  child { node [decision={$> 0.085$}{$\le 0.085$}] { 
  \texttt{Invasomes.Intensity\_UpperQuartile-\\Intensity\_Corr1Pathogen} }
    child { node [outcome] { infected } }
    child { node [decision={$> 12.00$}{$\le 12.00$}] { 
    \texttt{Invasomes.AreaShape\_Area} }
      child { node [decision={$< 1000$}{$\ge 1000$}] { 
        \texttt{Invasomes.AreaShape\_Area} }
        child { node [decision={$< 0.060$}{$\ge 0.060$}] { 
          \texttt{Invasomes.RadialDistribution\_FracAtD\_\\Corr1Actin\_1} }
          child { node [outcome] { infected } }
          child { node [outcome] { not infected } }
        }
        child { node [outcome] { not infected } }
      }
      child { node [outcome] { not infected } }
    }
  }
};

\end{tikzpicture}
\caption[Decision tree for \textit{Bartonella} infection scoring.]{Decision tree for \textit{Bartonella} infection scoring. In order to detect bona-fide invasomes, the first three or-linked decisions assemble a list of candidates while the following three and-linked decisions discard some erroneously included instances. In order to obtain the desired cell-based infection score, invasomes are mapped to cellular objects in a subsequent step.}
\end{figure}
